File: Plan Graph (sgware)/src/edu/uno/ai/planning/pg/Mutexes.java
package edu.uno.ai.planning.pg;

import java.util.HashMap;

class Mutexes {
	
	public static final int ALWAYS = -1;

	private final HashMap<Node, Integer> nodes = new HashMap<>();
	
	public void add(Node node, int level) {
		Integer end = nodes.get(node);
		if(end == null)
			nodes.put(node, level);
		else if(end == ALWAYS)
			return;
		else if(end < level)
			nodes.put(node, level);
	}
	
	public boolean contains(Node node, int level) {
		if(!node.exists(level))
			throw new IllegalArgumentException(node + " does not exist at level " + level + ".");
		Integer end = nodes.get(node);
		if(end == null)
			return false;
		else if(end == ALWAYS)
			return true;
		else
			return level <= end;
	}
	
	public void clear() {
		nodes.clear();
	}
}

File: UnitySokoban/Assets/Scripts/Planning/PlanGraphSGW/Mutexes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PlanGraphSGW
{
    class Mutexes
    {
        public static readonly int ALWAYS = -1;
        private readonly Dictionary<Node, int> nodes = new Dictionary<Node, int>();

        public void add(Node node, int level)
        {
            if (!nodes.ContainsKey(node))
                nodes[node] = level;
            else if (nodes[node] == ALWAYS)
                return;
            else if (nodes[node] < level)
                nodes[node] = level;
        }

        public bool contains(Node node, int level)
        {
            if (!node.exists(level))
                throw new ArgumentException(node + " does not exist at level " + level + ".");
            if (!nodes.ContainsKey(node))
                return false;
            else if (nodes[node] == ALWAYS)
                return true;
            else
                return level <= nodes[node];
        }

        public void clear()
        {
            nodes.Clear();
        }
    }

}
